packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1157,127 +1157,242 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
-          continue;
+      } on ToolExit catch (e) {
+        if (e.message != null) {
+          globals.printError(e.message!);
         }
-      }
-      // Otherwise, require an explicit choice.
-      if (candidates.length > 1) {
-        globals.printError(
-            'Plugin ${entry.key}:$platformKey has multiple possible implementations:\n'
-                '${candidates.map((Plugin plugin) => '  ${plugin.name}\n').join()}'
-                'To fix this issue, add one of these dependencies to pubspec.yaml.\n'
-        );
         hasResolutionError = true;
-        continue;
       }
     }
 
-    finalResolution.addAll(
-      pluginResolution.map((Plugin plugin) =>
+    pluginResolutions.addAll(
+      pluginResolution.values.map((Plugin plugin) =>
           PluginInterfaceResolution(plugin: plugin, platform: platformKey)),
     );
   }
+  if (hasPluginPubspecError) {
+    throwToolExit('Please resolve the plugin pubspec errors');
+  }
   if (hasResolutionError) {
     throwToolExit('Please resolve the plugin implementation selection errors');
   }
-  return finalResolution;
+  return pluginResolutions;
+}
+
+/// Validates conflicting plugin parameters in pubspec, such as
+/// `dartPluginClass`, `default_package` and `implements`.
+void _validatePlugin(Plugin plugin, String platformKey) {
+  final String? implementsPackage = plugin.implementsPackage;
+  final String? defaultImplPluginName = plugin.defaultPackagePlatforms[platformKey];
+
+  if (plugin.name == implementsPackage &&
+      plugin.name == defaultImplPluginName) {
+    // Allow self implementing and self as platform default.
+    return;
+  }
+
+  if (defaultImplPluginName != null) {
+    if (implementsPackage != null && implementsPackage.isNotEmpty) {
+      throwToolExit(
+          'Plugin ${plugin.name}:$platformKey which provides an implementation for $implementsPackage '
+          'can not also reference a default implementation for $defaultImplPluginName. '
+          'Ask the maintainers of ${plugin.name} to either remove the implementation via `implements: $implementsPackage` '
+          'or avoid referencing a default implementation via `platforms: $platformKey: default_package: $defaultImplPluginName`.\n');
+    }
+
+    if (_hasPluginInlineDartImpl(plugin, platformKey)) {
+      throwToolExit(
+          'Plugin ${plugin.name}:$platformKey which provides an inline implementation '
+          'can not also reference a default implementation for $defaultImplPluginName. '
+          'Ask the maintainers of ${plugin.name} to either remove the implementation via `platforms: $platformKey: dartPluginClass` '
+          'or avoid referencing a default implementation via `platforms: $platformKey: default_package: $defaultImplPluginName`.\n');
+    }
+  }
+}
+
+/// Determine if this [plugin] serves as implementation for an app-facing
+/// package for the given platform [platformKey].
+/// If so, return the package name, which the [plugin] implements.
```

Comment:
Blank line above this, per Effective Dart style.

Reply:
Done in 1f9cc0939880613363e5573b360524f4ccf38081